QuizApp/Presenters/Quizzes/QuizzesPresenter.swift

```diff
@@ -0,0 +1,90 @@
+//
+//  QuizzesPresenter.swift
+//  QuizApp
+//
+//  Created by Luka Namačinski on 12.05.2021..
+//
+
+import Foundation
+
+class QuizzesPresenter: QuizzesPresenterProtocol {
+    weak var coordinator: QuizzesCoordinator?
+    var dataService: NetworkServiceProtocol = NetworkService()
+    var delegate: QuizzesDelegate
+    
+    var quizzes: [Quiz] {
+        didSet {
+            if quizzes.count > 0 {
+                delegate.showQuizzes()
+            }
+            else {
+                delegate.showError()
+            }
+        }
+    }
+    var categorisedQuizzes: [QuizCategory : [Quiz]]
+    
+    init(delegate qd: QuizzesDelegate, coordinator qc: QuizzesCoordinator) {
+        self.delegate = qd
+        self.coordinator = qc
+        self.quizzes = []
+        self.categorisedQuizzes = [:]
+    }
+    
+    func fetchQuizzes() {
+        DispatchQueue.global().async {
+            self.dataService.fetchQuizes(completionHandler: {
```

Comment:
Svaki closure koji se asinkrono poziva bi trebao koristiti weak self, zbog toga što se taj closure može pozvati u trenutku kada self više ne postoji i u tom slučaju bi ti aplikacija pucala na tim mjestima.

```
            dataService.fetchQuizes(completionHandler: { [weak self]
                (result: Result<QuizCollection, RequestError>) -> Void in
                guard let self = self else { return }

                DispatchQueue.main.async {
                    switch result {
                    case .failure(let error):
                        self.delegate.setErrorMessage(message: error.localizedDescription)
                        self.quizzes = []
                    case .success(let fetchedCollection):
                        self.quizzes = fetchedCollection.quizzes
                        self.setFunFact()
                    }
                    self.categorisedQuizzes = Dictionary(grouping: self.quizzes, by: {$0.category})
                    self.delegate.reloadTable()
                }
            })
```